program/src/processor.rs

```diff
@@ -188,20 +189,177 @@ impl Processor {
         mango_account.owner = *owner_ai.key;
         mango_account.order_market = [FREE_ORDER_SLOT; MAX_PERP_OPEN_ORDERS];
         mango_account.meta_data = MetaData::new(DataType::MangoAccount, 0, true);
+        mango_account.not_upgradable = true;
+        Ok(())
+    }
+
+    #[inline(never)]
+    fn close_mango_account(program_id: &Pubkey, accounts: &[AccountInfo]) -> MangoResult<()> {
+        const NUM_FIXED: usize = 3;
+        let accounts = array_ref![accounts, 0, NUM_FIXED];
+
+        let [
+            mango_group_ai,     // write
+            mango_account_ai,   // write
+            owner_ai,           // write, signer
+        ] = accounts;
+
+        check_eq!(&mango_account_ai.owner, &program_id, MangoErrorCode::InvalidOwner)?;
+        let mut mango_account =
+            MangoAccount::load_mut_checked(mango_account_ai, program_id, &mango_group_ai.key)?;
+        check_eq!(&mango_account.owner, owner_ai.key, MangoErrorCode::InvalidOwner)?;
+        check!(owner_ai.is_signer, MangoErrorCode::InvalidSignerKey)?;
+        check!(mango_account.meta_data.version > 0, MangoErrorCode::InvalidAccountState)?;
+
+        // Check deposits and borrows are zero
+        for i in 0..MAX_TOKENS {
+            check_eq!(mango_account.deposits[i], ZERO_I80F48, MangoErrorCode::InvalidAccountState)?;
+            check_eq!(mango_account.borrows[i], ZERO_I80F48, MangoErrorCode::InvalidAccountState)?;
+        }
+        // Check no perp positions or orders
+        for perp_account in mango_account.perp_accounts.iter() {
+            check_eq!(perp_account.base_position, 0, MangoErrorCode::InvalidAccountState)?;
+            check_eq!(
+                perp_account.quote_position,
+                ZERO_I80F48,
+                MangoErrorCode::InvalidAccountState
+            )?;
+
+            check!(perp_account.is_liquidatable(), MangoErrorCode::InvalidAccountState)?;
+        }
+        // Check no msrm
+        check_eq!(mango_account.msrm_amount, 0, MangoErrorCode::Default)?;
+        // Check not being liquidated/bankrupt
+        check!(!mango_account.being_liquidated, MangoErrorCode::BeingLiquidated)?;
+        check!(!mango_account.is_bankrupt, MangoErrorCode::Bankrupt)?;
+
+        // Check open orders accounts closed
+        for open_orders_key in mango_account.spot_open_orders.iter() {
+            check_eq!(open_orders_key, &Pubkey::default(), MangoErrorCode::Default)?;
+        }
+        // Check advanced orders account closed
+        check_eq!(&mango_account.advanced_orders_key, &Pubkey::default(), MangoErrorCode::Default)?;
+
+        // Transfer lamports to owner
+        program_transfer_lamports(mango_account_ai, owner_ai, mango_account_ai.lamports())?;
+
+        let mut mango_group = MangoGroup::load_mut_checked(mango_group_ai, program_id)?;
+        mango_group.num_mango_accounts -= 1;
+
+        // Prevent account being loaded by program
+        mango_account.meta_data.is_initialized = false;
```

Comment:
TBH, `is_initialized` going `false` is a bit scary. I think some code may rely on the fact that new accounts are zero-initialized when freshly created. Could we write 0 all over the account data or something, to avoid people quickly re-initing the closed account and somehow retaining some old data?

Reply:
Most of the account properties are already in their default state if the program gets to this point, I will make sure the remaining properties are reset too